### PR TITLE
`open`, `rm`, `umv`, `cp`, `rm` and `du`: Don't globs if inputs are variables or string interpolation

### PR DESCRIPTION
**Pull request by WindSoilder**
# Description
This is a follow up to https://github.com/nushell/nushell/pull/11621#issuecomment-1937484322

Also Fixes: #11838 

## About the code change
It applys the same logic when we pass variables to external commands:

https://github.com/nushell/nushell/blob/0487e9ffcbc57c2d5feca606e10c3f8221ff5e00/crates/nu-command/src/system/run_external.rs#L162-L170

That is: if user input dynamic things(like variables, sub-expression, or string interpolation), it returns a quoted `NuPath`, then user input won't be globbed
 
# User-Facing Changes
Given two input files: `a*c.txt`, `abc.txt`

* `let f = "a*c.txt"; rm $f` will remove one file: `a*c.txt`. 
~* `let f = "a*c.txt"; rm --glob $f` will remove `a*c.txt` and `abc.txt`~
* `let f: glob = "a*c.txt"; rm $f` will remove `a*c.txt` and `abc.txt`

## Rules about globbing with *variable*
Given two files: `a*c.txt`, `abc.txt`
| Cmd Type | example | Result |
| ----- | ------------------ | ------ |
| builtin | let f = "a*c.txt"; rm $f | remove `a*c.txt` |
| builtin | let f: glob = "a*c.txt"; rm $f | remove `a*c.txt` and `abc.txt` 
| builtin | let f = "a*c.txt"; rm ($f \| into glob) | remove `a*c.txt` and `abc.txt`
| custom | def crm [f: glob] { rm $f }; let f = "a*c.txt"; crm $f | remove `a*c.txt` and `abc.txt`
| custom | def crm [f: glob] { rm ($f \| into string) }; let f = "a*c.txt"; crm $f | remove `a*c.txt`
| custom | def crm [f: string] { rm $f }; let f = "a*c.txt"; crm $f | remove `a*c.txt`
| custom | def crm [f: string] { rm $f }; let f = "a*c.txt"; crm ($f \| into glob) | remove `a*c.txt` and `abc.txt`

In general, if a variable is annotated with `glob` type, nushell will expand glob pattern.  Or else, we need to use `into | glob` to expand glob pattern

# Tests + Formatting
Done

# After Submitting
I think `str glob-escape` command will be no-longer required.  We can remove it.

### CODE REVIEW

**Comment by kubouch** [comment]
I think it is a good change to remove auto-globbing from variables and string interpolations. Also agreed that we can get rid of `str glob-escape`.

I'm just not sold on adding `--glob` to glob-capable builtin commands because it seems like a limited solution that does not solve the problem in all cases where globs can occur:

First, custom commands:
```
> def gcmd [g: glob] { open $g }
> 'a' > a.txt
> 'b' > b.txt
> gcmd *.txt 
# opens a.txt and b.txt
> let g = '*.txt'
> gcmd $g  # <-- How to expand the glob here?
```
_Note: `gcmd *.txt` doesn't work on this PR but works on main, I think it's a bug in this PR._

Second, externals. Currently, we do expand globs for externals:
```
> mkdir cliargs
> cd cliargs
> cargo init
> 'fn main() {
    println!("{:?}", std::env::args().collect::<Vec<String>>());
}' o> src/main.rs
> let gl = '*'
> cargo run -- $gl
["/home/kubouch/.cargo/target/debug/cliargs", ".git", ".gitignore", "Cargo.lock", "Cargo.toml", "src"]
```
If we remove the variable/string interp. auto-globbing (which we should), there would be no way to allow it for externals because we can't add a `--glob` flag there.

Third, plugins? Any other cases?

I am thinking of three alternatives (not mutually exclusive):
1. Add a new `Value::Glob` variant which could be passed around and cast to/from string (`into string`, `into glob`) to prevent glob expansion or `collect`ed into a list if needed. Example:
```
> let g: glob = 'a*c.txt'
> rm $g  # removes abc.txt, azc.txt, ...
> rm a*c.txt   # same
> rm ($g | into string) # removes only a*c.txt
> rm 'a*c.txt' # same
> external-command $g # passes [abc.txt, azc.txt, ...] to the external
> external-command ($g | into string) # passes only [a*c.txt] to the external
```
2. Not adding a new value type, but rely more heavily on the spread operator:
```
> let g: string = 'a*c.txt'
> rm ...(glob $g)  # removes abc.txt, azc.txt, ...
> rm a*c.txt   # same
> rm $g # removes only a*c.txt
> rm 'a*c.txt' # same
> external-command ...(glob $g)  # passes [abc.txt, azc.txt, ...] to the external
> external-command $g  # passes only [a*c.txt] to the external
```
3. Allow casting of lists to globs:
```
> def gcmd [g: glob] { open $g }
> let g = '*.txt'
> gcmd (glob $g)  # opens a.txt and b.txt
```

We have two other features related to globbing: `Value::QuotedString` and the `glob` command. So whatever we do, these should be kept in mind so that we don't duplicate functionality. In the option 1, for example, we might be able to remove both of them.

**Comment by fdncred** [comment]
A much more radical approach to number 2 of @kubouch's comments above is to remove globbing from every command and only allow globs via the glob command and the used with the spread operator. It could cause people to lose their mind, but it would simplify things.

I'm up for giving most anything a try in order to get something nushell-y that we love.

**Comment by WindSoilder** [comment]
> 1. Add a new Value::Glob variant which could be passed around and cast to/from string (into string, into glob) to prevent glob expansion or collected into a list if needed. Example:

I'm not really sure whether we can implement it in this way.  Because the value are directed by `SyntaxShape`, that says, it's really hard to implement this: `rm ($g | into string) # removes only a*c.txt`, because `($g | into string)` will always be glob in `rm`.

Personally I like 2nd and 3rd solution :)

File: crates/nu-command/src/conversions/into/glob.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        _ => operate(action, args, input, head, engine_state.ctrlc.clone()),
+    }
+}
+
+fn action(input: &Value, _args: &Arguments, span: Span) -> Value {
+    match input {
+        Value::String { val, .. } => Value::glob(val.to_string(), false, span),
+        x => Value::error(
+            ShellError::CantConvert {
+                to_type: String::from("cawtring"),
```
Comment (fdncred) [inline crates/nu-command/src/conversions/into/glob.rs:113]: cawtring? 😆

**Comment by WindSoilder** [comment]
I still fail into a trouble to play with the following two cases, given two files: `touch a*c.txt abc.txt`.

|  Number | Custom command  |  run |  result
| ------- | ------------ | -------- | ------ | 
| 1 | `def foo [g: glob] { rm $g }` | `let f = "a*c.txt"; foo $f` | `a*c.txt` and `abc.txt` will be removed |
| 2 | `def foo [g: string] { rm $g }` | `let f = "a*c.txt"; foo $f` | only `a*c.txt` is removed |

* for 1st one: there is no way to disable glob, because argument `g` is always glob type.
* for 2nd one: to apply a glob, we need to apply `into glob` command: `def foo [g: string] { rm ($g | into glob) }`

Not really sure if the behavior is ok

File: crates/nu-command/src/conversions/into/glob.rs (hunk 1 above)
Comment (WindSoilder) [reply crates/nu-command/src/conversions/into/glob.rs]: Oh..My bad

**Comment by WindSoilder** [comment]
Reply to @kubouch 's proposal:
> Add a new Value::Glob variant which could be passed around and cast to/from string (into string, into glob) to prevent glob expansion or collected into a list if needed.

Done :)

>  Not adding a new value type, but rely more heavily on the spread operator

Nushell supports this feature on most of filesystem command except `ls` and `du`.  Because they don't accept rest parameters.

> Allow casting of lists to globs:  def gcmd [g: glob] { open $g }; let g = '*.txt'; gcmd (glob $g)  # opens a.txt and b.txt

Currently this pr doesn't implement it.  We can do this by `let g: glob = "*.txt"; gcmd $g` or `let g = "*.txt"; gcmd ($g | into glob)`

**Comment by WindSoilder** [comment]
I think we can land and iterate through it

**Comment by matklad** [comment]
> `let f: glob = "a*c.txt"; rm $f` will remove a*c.txt and abc.txt

That's a _fantastic_ solution to the problem, thanks!

File: crates/nu-command/src/conversions/into/glob.rs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    fn examples(&self) -> Vec<Example> {
+        vec![
+            Example {
+                description: "convert string to glob",
+                example: "'1234' | into glob",
+                result: Some(Value::test_string("1234")),
+            },
+            Example {
+                description: "convert filepath to string",
```
Comment (kubouch) [inline crates/nu-command/src/conversions/into/glob.rs:74]: I think this should say "to glob" instead of "to string", right?

Comment (WindSoilder) [reply crates/nu-command/src/conversions/into/glob.rs:74]: Yup.  It's change by #11992

**Pull request merged**